connector/signaltometrics/connector.go

```diff
@@ -0,0 +1,281 @@
+					case pmetric.MetricTypeEmpty:
+						multiError = errors.Join(multiError, fmt.Errorf("metric %q: invalid metric type: %v", metric.Name(), metric.Type()))
+					}
+				}
+			}
+		}
+	}
+	if multiError != nil {
+		return multiError
+	}
+	aggregator.Finalize(sm.dpMetricDefs)
+	return sm.next.ConsumeMetrics(ctx, processedMetrics)
+}
+
+func (sm *signalToMetrics) ConsumeLogs(ctx context.Context, logs plog.Logs) error {
+	var multiError error
+	processedMetrics := pmetric.NewMetrics()
+	processedMetrics.ResourceMetrics().EnsureCapacity(logs.ResourceLogs().Len())
+	aggregator := aggregator.NewAggregator[ottllog.TransformContext](processedMetrics)
+	for i := 0; i < logs.ResourceLogs().Len(); i++ {
+		resourceLog := logs.ResourceLogs().At(i)
+		resourceAttrs := resourceLog.Resource().Attributes()
+		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
+			scopeLog := resourceLog.ScopeLogs().At(j)
+			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
+				log := scopeLog.LogRecords().At(k)
+				logAttrs := log.Attributes()
+				for _, md := range sm.logMetricDefs {
+					filteredLogAttrs := getFilteredAttributes(logAttrs, md.Attributes)
+					if filteredLogAttrs.Len() != len(md.Attributes) {
+						// If all the configured attributes are not present in
+						// source metric then don't count them.
+						continue
+					}
+					var filteredResAttrs pcommon.Map
+					if len(md.IncludeResourceAttributes) > 0 {
+						filteredResAttrs = getFilteredAttributes(resourceAttrs, md.IncludeResourceAttributes)
+					} else {
+						// Copy resource attrs to avoid mutating data
+						filteredResAttrs = pcommon.NewMap()
+						resourceAttrs.CopyTo(filteredResAttrs)
+					}
+					tCtx := ottllog.NewTransformContext(log, scopeLog.Scope(), resourceLog.Resource(), scopeLog, resourceLog)
+					multiError = errors.Join(multiError, aggregator.Aggregate(ctx, tCtx, md, filteredResAttrs, filteredLogAttrs, 1))
+				}
+			}
+		}
+	}
+	if multiError != nil {
+		return multiError
+	}
+	aggregator.Finalize(sm.logMetricDefs)
+	return sm.next.ConsumeMetrics(ctx, processedMetrics)
+}
+
+// calculateAdjustedCount calculates the adjusted count which represents
+// the number of spans in the population that are represented by the
+// individually sampled span. If the span is not-sampled OR if a non-
+// probability sampler is used then adjusted count defaults to 1.
+// https://github.com/open-telemetry/oteps/blob/main/text/trace/0235-sampling-threshold-in-trace-state.md
+func calculateAdjustedCount(tracestate string) int64 {
+	w3cTraceState, err := sampling.NewW3CTraceState(tracestate)
+	if err != nil {
+		return 1
+	}
+	otTraceState := w3cTraceState.OTelValue()
+	if otTraceState == nil {
+		return 1
+	}
+	if len(otTraceState.TValue()) == 0 {
+		// For non-probabilistic sampler OR always sampling threshold, default to 1
+		return 1
+	}
+	// TODO (lahsivjar): Handle fractional adjusted count. One way to do this
+	// would be to scale the values in the histograms for some precision.
```

Comment:
Another option would be to apply some kind of probabilistic rounding. For example, if the adjusted count is 2.3, you return 2 with a 60% chance and 3 with a 30% chance.

Reply:
Probabilistic rounding is a good idea. I will create an issue for handling this and add the ideas there.

Reply:
Created an issue: https://github.com/elastic/opentelemetry-collector-components/issues/170